### PR TITLE
100% readonly fields and ReadOnlySpan

### PR DESCRIPTION
**Pull request by Jehoel**
I'll admit this PR is a bit busy, I had a bad case of "one more thing...".

------------

My main objective was to change `class Hashids` such that all of its fields were `readonly`, as currently the (somewhat twisted) initialization logic for `_alphabet` and `_seps` happens in a separate non-static method which mutates `this` fields, which isn't permitted by `readonly`.

The main commit that matters is c0564ad10a287223be44ef49d9a6916cf2fba398 which moves that init logic into a new `static void InitCharArrays`.

------------

Another significant change is in 99bd19fb3795a11a427481f65381678faddbd96a which replaces occurrences of `char[]` and `long[]` with `ReadOnlySpan<>` to signal intent that a parameter is not mutated, which is always a concern when passing around arrays as parameters (given C# lacks C++-style `const`-correctness or first-class support for immutability).

For compatibility with pre-.NET Standard 2.1 consumers (.NET Framework, etc) the library includes a thin `internal struct ReadOnlySpan<T>` reimplementation, which _should_ be zero-cost (so long as it isn't moved to the GC heap or boxed), and the JIT _should_ also inline all or most of its members.

I actually didn't put much (any?) effort into other optimizations, though I'm happy with the benchmark results below.

I'm tempted to add another commit with C# 8.0 nullability too...

**UPDATE:** I successfully implemented C# `<Nullable>enabled</Nullable>` as well as other performance improvements (such as eliminating array allocations in the various `Encode`/`Decode` overloads), but things snowballed and it became _far too much_ to put into a PR, so I've moved those changes to a separate branch I'll submit a PR for only after this is approved (and hopefully merged).

------------

## Benchmark results:

* For .NET Framework 4.8 there is an insignificant performance cost, I assume from the extra indirection from the `ReadOnlySpan` implementation. I don't believe that the .NET 4.8 runtime and JIT has any optimizations for `readonly` fields though.
* For .NET 5.0 there's an insignificant performance boost, which I assume is from `readonly` fields and the fact that `ReadOnlySpan` is heavily optimized in .NET Core.

#### Before

|         Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
|  RoundtripInts | 4.316 us | 0.0326 us | 0.0272 us | 0.1297 |     - |     - |     859 B |
| RoundtripLongs | 5.077 us | 0.0193 us | 0.0161 us | 0.1450 |     - |     - |     971 B |
|   RoundtripHex | 5.418 us | 0.0225 us | 0.0210 us | 0.3052 |     - |     - |    1966 B |

#### After

|         Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
|  RoundtripInts | 4.321 us | 0.0360 us | 0.0337 us | 0.1297 |     - |     - |     859 B |
| RoundtripLongs | 5.129 us | 0.0357 us | 0.0316 us | 0.1450 |     - |     - |     971 B |
|   RoundtripHex | 5.350 us | 0.0389 us | 0.0364 us | 0.3052 |     - |     - |    1966 B |

#### Before

|         Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
|  RoundtripInts | 3.949 us | 0.0262 us | 0.0245 us | 0.0687 |     - |     - |     576 B |
| RoundtripLongs | 4.515 us | 0.0198 us | 0.0185 us | 0.0687 |     - |     - |     576 B |
|   RoundtripHex | 3.783 us | 0.0118 us | 0.0092 us | 0.1793 |     - |     - |    1512 B |

#### After

|         Method |     Mean |     Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|----------:|----------:|-------:|------:|------:|----------:|
|  RoundtripInts | 3.910 us | 0.0134 us | 0.0125 us | 0.0610 |     - |     - |     512 B |
| RoundtripLongs | 4.461 us | 0.0200 us | 0.0187 us | 0.0610 |     - |     - |     512 B |
|   RoundtripHex | 3.774 us | 0.0153 us | 0.0136 us | 0.1602 |     - |     - |    1344 B |

### CODE REVIEW

File: src/Hashids.net/Hashids.cs [hunk 1]
```diff
@@ -350,16 +368,12 @@ private long[] GetNumbersFrom(string hash)
...
-            {
-                i = 1;
-            }
+            var i = (hashArray.Length is 3 or 2 ) ? 1 : 0;
 
             var hashBreakdown = hashArray[i];
             var lottery = hashBreakdown[0];
 
-            if (lottery == default(char))
+            if (lottery == '\0') /* default(char) == '\0' */
```
Comment (manigandham) [inline src/Hashids.net/Hashids.cs:376]: [Default value expressions are also constant expressions](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#default-value-expressions) for value types in C#, which are evaluated at compile-time already so there's no performance advantage from this.

It's better to stick with `default(char)` for readability.

**Review by manigandham — CHANGES_REQUESTED** [review]

**Comment by manigandham** [comment]
Hey @Jehoel 
Thanks for this, finally got a chance to review it all and this is some great work!

One small change noted there but I'll go ahead and merge to get this in. Looking forward to the other changes, but please keep the PRs as small as you can. Easier to review multiple PRs if needed.

Thanks

File: src/Hashids.net/Hashids.cs (hunk 1 above)
Comment (manigandham) [reply src/Hashids.net/Hashids.cs:376]: @Jehoel I'll fix this later

Comment (Jehoel) [reply src/Hashids.net/Hashids.cs:376]: Yes, I know `default` is a constant-expression, the reason I opted for `'\0'` instead is to signify intent: as `default` is often used in C# with the connotation that the value is unimportant, whereas in this case the zero value is _very significant_ to the algorithm so I wanted to be explicit.

**Comment by Jehoel** [comment]
@manigandham Uhhh, I wasn't prepared for it to be merged as-is - there's a bunch of TODOs and Questions left in comments....

**Comment by manigandham** [comment]
@Jehoel 

I only see 3 TODOs; - One about short variable names and the others about the seps/alphabet characters. We can take care of that after your other changes before doing a release.

This project was ported directly from the Javascript version originally so it followed the same layout and naming. Feel free to create a new branch or just push to this one and we can make another PR.

**Pull request merged**